### PR TITLE
Monomorphization of Simplifier, leg #1

### LINKED ISSUE #3325: Serokell: monomorphisation + specialization

### PR DESCRIPTION
**Pull request by breakerzirconia**
Partially fixes #3325

### Scope:

The 1st leg of the Simplifier monomorphization.
Modules affected:
* `Kore.Reachability.Prove`
* `Kore.Rewrite.RewriteStep`
* `Kore.Rewrite.Step`
* `Kore.Simplify.Application`

### Estimate:

---

###### Review checklist

The author performs the actions on the checklist. The reviewer evaluates the work and checks the boxes as they are completed.

-   [ ] **Summary.** Write a summary of the changes. Explain what you did to fix the issue, and why you did it. Present the changes in a logical order. Instead of writing a summary in the pull request, you may push a clean Git history.
-   [ ] **Documentation.** Write documentation for new functions. Update documentation for functions that changed, or complete documentation where it is missing.
-   [ ] **Tests.** Write unit tests for every change. Write the unit tests that were missing before the changes. Include any examples from the reported issue as integration tests.
-   [ ] **Clean up.** The changes are already clean. Clean up anything near the changes that you noticed while working. This does not mean only spatially near the changes, but logically near: any code that interacts with the changes!

### CODE REVIEW

**Comment by breakerzirconia** [comment]
[report-breakerzircon--leg-1_master.pdf](https://github.com/runtimeverification/haskell-backend/files/9928217/report-breakerzircon--leg-1_master.pdf)

The performance test was run against two branches:
`master`: 0cc2eb046101cba0080b63e04a5b5a0f1374a99f
`breakerzirconia/monomorphize-simplifier-leg-1`: 08aac3df14fb6ae61567b889de642df6a44c833e
Result: (100% / 89.46% - 1) * 100% = **11.78%** performance improvement.

**Comment by ana-pantilie** [comment]
Woo! Nice results! 🎉 

I wonder why those two tests are slower on this branch. As a follow-up, please profile them with this change, and let us know if anything stands out. The two tests are:
- bug approve succes
- bug vat mului pas

File: kore/src/Kore/Rewrite/RewriteStep.hs [hunk 1]
```diff
@@ -141,14 +138,13 @@ finalizeAppliedRule
...
 constructConfiguration ::
-    MonadSimplify simplifier =>
     -- | SideCondition containing metadata
     SideCondition RewritingVariableName ->
     -- | Applied condition
     Condition RewritingVariableName ->
     -- | Final configuration
     Pattern RewritingVariableName ->
-    LogicT simplifier (Pattern RewritingVariableName)
+    LogicT (LogicT Simplifier) (Pattern RewritingVariableName)
```
Comment (ana-pantilie) [inline kore/src/Kore/Rewrite/RewriteStep.hs:147]: Interesting that the `LogicT` is nested here. Can you try to flatten this (in a separate PR) as an experiment?

**Review by ana-pantilie — APPROVED** [review]

**Comment by breakerzirconia** [comment]
> Woo! Nice results! 🎉
> 
> I wonder why those two tests are slower on this branch. As a follow-up, please profile them with this change, and let us know if anything stands out. The two tests are:
> 
>     * bug approve succes
> 
>     * bug vat mului pas

Thank you, will do! 
Also, my apologies, I messed up the percentages a little bit. Since the original number was 100%, and the resulting is 89.46%, there is 100% - 89.46% = **10.54%** performance improvement.

**Pull request merged**